src/util.rs

```diff
@@ -7,149 +7,44 @@
 
 //! Helper functions
 
-use core::array::IntoIter;
+use core::convert::TryFrom;
 
-use generic_array::typenum::U0;
+use generic_array::typenum::{IsLess, U2, U256};
 use generic_array::{ArrayLength, GenericArray};
 
 use crate::{Error, Result};
 
-// Corresponds to the I2OSP() function from RFC8017
-pub(crate) fn i2osp<L: ArrayLength<u8>>(input: usize) -> Result<GenericArray<u8, L>> {
-    const SIZEOF_USIZE: usize = core::mem::size_of::<usize>();
-
-    // Make sure input fits in output.
-    if (SIZEOF_USIZE as u32 - input.leading_zeros() / 8) > L::U32 {
-        return Err(Error::SerializationError);
-    }
-
-    let mut output = GenericArray::default();
-    output[L::USIZE.saturating_sub(SIZEOF_USIZE)..]
-        .copy_from_slice(&input.to_be_bytes()[SIZEOF_USIZE.saturating_sub(L::USIZE)..]);
-    Ok(output)
-}
-
-/// Computes `I2OSP(len(input), max_bytes) || input` and helps hold output
-/// without allocation.
-pub(crate) struct Serialize<'a, L1: ArrayLength<u8>, L2: ArrayLength<u8> = U0> {
-    octet: GenericArray<u8, L1>,
-    input: Input<'a, L2>,
-}
-
-enum Input<'a, L: ArrayLength<u8>> {
-    Owned(GenericArray<u8, L>),
-    Borrowed(&'a [u8]),
-}
-
-impl<'a, L1: ArrayLength<u8>, L2: ArrayLength<u8>> IntoIterator for &'a Serialize<'a, L1, L2> {
-    type Item = &'a [u8];
-
-    type IntoIter = IntoIter<&'a [u8], 2>;
-
-    fn into_iter(self) -> Self::IntoIter {
-        // MSRV: array `into_iter` isn't available in 1.51
-        #[allow(deprecated)]
-        IntoIter::new([
-            &self.octet,
-            match self.input {
-                Input::Owned(ref bytes) => bytes,
-                Input::Borrowed(bytes) => bytes,
-            },
-        ])
-    }
-}
-
-impl<'a, L1: ArrayLength<u8>, L2: ArrayLength<u8>> Serialize<'a, L1, L2> {
-    // Variation of `serialize` that takes a borrowed `input.
-    pub(crate) fn from(input: &[u8]) -> Result<Serialize<L1>> {
-        Ok(Serialize {
-            octet: i2osp::<L1>(input.len())?,
-            input: Input::Borrowed(input),
-        })
-    }
-
-    pub(crate) fn from_owned(input: GenericArray<u8, L2>) -> Result<Serialize<'static, L1, L2>> {
-        Ok(Serialize {
-            octet: i2osp::<L1>(input.len())?,
-            input: Input::Owned(input),
-        })
-    }
-}
-
-macro_rules! chain_name {
-    ($var:ident, $mod:ident) => {
-        $mod
-    };
-    ($var:ident) => {
-        $var
-    };
-}
-
-macro_rules! chain_skip {
-    ($var:ident, $feed:expr) => {
-        $feed
-    };
-    ($var:ident) => {
-        &$var
-    };
+pub(crate) fn i2osp_2(input: usize) -> Result<GenericArray<u8, U2>> {
```

Comment:
Why is this named `i2osp_2`? Do you perhaps just mean `i2osp`?

Reply:
I named this `i2osp_2` because `i2osp` was still alive and this one wasn't generic, it simply always used a size of `2`. Later on it just happened that all instances of `i2osp` were replaced with `i2osp_2` so it was left like this.

I think the naming is still appropriate as the official `i2osp` function takes a length, that one doesn't, it always uses a length of `2`, which is the only thing needed in VOPRF.